app/src/org/commcare/utils/DeviceIdentifier.java

```diff
@@ -0,0 +1,41 @@
+package org.commcare.utils;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+import androidx.preference.PreferenceManager;
+import java.util.UUID;
+
+/**
+ * A Utility to get a unique identifier for app's installation.
+ * @author $|-|!˅@M
+ */
+public final class DeviceIdentifier {
+
+    private static volatile String uuid;
+    private static final String PREFIX = "commcare_";
+    private static final String KEY_DEVICE_IDENTIFIER = "commcare_device_id";
+
+    /**
+     * Returns a unique(very highly likely) identifier for this device.
+     * This identifier is prefixed with "commcare_"
+     *
+     * This id may change when user re-installs the app or does a factory reset.
+     */
+    public static String getDeviceIdentifier(Context context) {
+        if (uuid == null) {
```

Comment:
there are 2 nested if here with `uuid == null` check. Is there a particular reason behind that ?

Reply:
Yeah, this pattern is also popular as _double-checked locking idiom_, used to reduce the overhead of acquiring a lock by first testing the locking criterion without actually acquiring the lock.

Reply:
Interesting!